src/QsCompiler/Compiler/Templates/QirDriverCpp.tt

```diff
@@ -1,16 +1,278 @@
+#include <memory>
+#include <vector>
 
-extern "C" void <#= entryPointOperation.Name #>(); // NOLINT
+#include "CLI11.hpp"
 
+#include "QirContext.hpp"
+#include "QirRuntime.hpp"
+#include "SimFactory.hpp"
+
+using namespace Microsoft::Quantum;
+using namespace std;
+    
+<# if (entryPointOperation.ContainsArgumentType(DataType.ArrayType)) { #>
+struct InteropArray
+{
+    int64_t Size;
+    void* Data;
+
+    InteropArray(int64_t size, void* data) :
+        Size(size),
+        Data(data){}
+};
+
+template<typename T>
+unique_ptr<InteropArray> CreateInteropArray(vector<T>& v)
+{
+    unique_ptr<InteropArray> array(new InteropArray(v.size(), v.data()));
+    return array;
+}
+
+template<typename S, typename D>
+void TranslateVector(vector<S>& sourceVector, vector<D>& destinationVector, function<D(S&)> translationFunction)
+{
+    destinationVector.resize(sourceVector.size());
+    transform(sourceVector.begin(), sourceVector.end(), destinationVector.begin(), translationFunction);
+
+<# } #>
+<# if (entryPointOperation.ContainsArgumentType(DataType.RangeType) || entryPointOperation.ContainsArrayType(DataType.RangeType)) { #>
+using RangeTuple = tuple<int64_t, int64_t, int64_t>;
+struct InteropRange
+{
+    int64_t Start;
+    int64_t Step;
+    int64_t End;
+
+    InteropRange() :
+        Start(0),
+        Step(0),
+        End(0){}
+
+    InteropRange(RangeTuple rangeTuple) :
+        Start(get<0>(rangeTuple)),
+        Step(get<1>(rangeTuple)),
+        End(get<2>(rangeTuple)){}
+};
+
+unique_ptr<InteropRange> CreateInteropRange(RangeTuple rangeTuple)
+{
+    unique_ptr<InteropRange> range(new InteropRange(rangeTuple));
+    return range;
+}
+
+InteropRange* TranslateRangeTupleToInteropRangePointer(RangeTuple& rangeTuple)
+{
+    InteropRange* range = new InteropRange(rangeTuple);
+    return range;
+}
+
+<# } #>
+<# if (entryPointOperation.ContainsArrayType(DataType.RangeType)) { #>
+template<typename T>
+void FreePointerVector(vector<T*>& v)
+{
+    for (auto p : v)
+    {
+        delete p;
+    }
+}
+<# } #>
+
+// This is the function corresponding to the QIR entry-point.
+extern "C" void <#= entryPointOperation.Name #>( // NOLINT
+<# for (int i = 0; i < entryPointOperation.InteropArguments.Count; i++) {
+    var arg = entryPointOperation.InteropArguments[i];
+    Write($"    {arg.CppType()} {arg.Name}");
+    if (i < entryPointOperation.InteropArguments.Count-1) {
+        WriteLine(",");
+    }
+} #>
+
+);
+
+<# if (entryPointOperation.ContainsArgumentType(DataType.BoolType) || entryPointOperation.ContainsArrayType(DataType.BoolType)) { #>
+const char InteropFalseAsChar = 0x0;
+const char InteropTrueAsChar = 0x1;
+map<string, bool> BoolAsCharMap{
+    {"0", InteropFalseAsChar},
+    {"false", InteropFalseAsChar},
+    {"1", InteropTrueAsChar},
+    {"true", InteropTrueAsChar}};
+
+<# } #>
+<# if (entryPointOperation.ContainsArgumentType(DataType.PauliType) || entryPointOperation.ContainsArrayType(DataType.PauliType)) { #>
```

Comment:
It would be good to group global variables and auxiliary functions that share the same condition. I think the only ones left to consolidate are the ones related to Pauli and Pauli arrays.

Reply:
done